A59-audit-logging.md

```diff
@@ -0,0 +1,281 @@
+[RBAC filter]: https://github.com/envoyproxy/envoy/blob/main/api/envoy/extensions/filters/http/rbac/v3/rbac.proto
+[RBAC policy]: https://github.com/envoyproxy/envoy/blob/main/api/envoy/config/rbac/v3/rbac.proto
+[HttpConnectionManager proto]: https://github.com/envoyproxy/envoy/blob/main/api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto
+
+### Related Proposals: 
+* [gRFC A41: xDS RBAC Support][A41]
+* [gRFC A43: gRPC Authorization API][A43]
+
+[A41]: A41-xds-rbac.md
+[A43]: A43-grpc-authorization-api.md
+
+## Proposal
+
+### Audit Context
+
+The table below lists the metadata we will include in the audit context,
+which will be made available to audit loggers during an audit event.
+
+It is worth noting that audit happens right after the authorization
+decision is made, so the status only reflects that but not the eventual
+one from the server. 
+
+|Fields             |Additional Information                                                             |Example                |
+|-------------------|-----------------------------------------------------------------------------------|-----------------------|
+|RPC Method         |Method the RPC hit.                                                                |"/pkg.service/foo"     |
+|Principal          |Identity the RPC. Currently only available in certificate-based TLS authentication.|"spiffe://foo/user1"   |
+|Timestamp          |Time when the server saw the RPC.                                                  |1649376211             |
+|Policy Name        |Name of the policy that invokes audit logging.                                     |"policy_name"          |
+|Authorized         |Boolean indicating whether the RPC is authorized or not.                           |true                   |
+
+### xDS API Changes
+
+xDS API changes are needed because gRPC authorization is entirely based on [xDS RBAC policy][RBAC policy].
+Moreoever, audit logging is a common feature that other xDS clients (namely Envoy) should benefit from as well.
+
+We will add an audit condition enum (see [PR](https://github.com/envoyproxy/envoy/pull/26001))
+in the [xDS RBAC policy][RBAC policy] as below:
+
+```
+package envoy.config.rbac.v3;
+
+message RBAC {
+  ...
+
+  // Deny and allow here refer to RBAC decisions, not actions.
+  enum AuditCondition {
+    // Never audit.
+    NONE = 0;
+
+    // Audit when RBAC denies the request.
+    ON_DENY = 1;
+
+    // Audit when RBAC allows the request.
+    ON_ALLOW = 2;
+
+    // Audit whether RBAC allows or denies the request.
+    ON_DENY_AND_ALLOW = 3;
+  }
+
+  // Condition for the audit logging to be happen, specific to HTTP filters.
+  // If condition is met, all the audit loggers configured in the HCM will be invoked.
+  //
+  AuditCondition audit_condition = 3;
+}
+```
+
+Note that `DENY` and `ALLOW` in the enum refer to the RBAC decisions, not
+RBAC actions. For example, we consider it as `DENY` decision when a RBAC
```

Comment:
This is a little confusing.  I would recommend instead:
For example, we consider it to be a `DENY` decision when a RBAC rejects the RPC, regardless of whether it is because there is no policy match from a RBAC with the `ALLOW` action, or because there is a match for a RBAC with a `DENY` action.  Likewise, it is an `ALLOW` decision when there are no matches to RBACs with a `DENY` action and at least one match to an RBAC with an `ALLOW` action.

Reply:
Thank you, Larry! I took your suggested piece.